README.md
# Genetic Algorithm

Implement Genetic Algorithm in JavaScript.


docs/CHANGELOG.md
## [1.0.6](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.5...v1.0.6) (2020-08-10)


### Bug Fixes

* complete the basic optimization ([206a55a](https://github.com/jeantimex/genetic-algorithm/commit/206a55ad6a9b09318d0a21e6aff29b4e0a28526f))

## [1.0.5](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.4...v1.0.5) (2020-08-09)


### Bug Fixes

* optimize the code phase 1 ([2df4d2c](https://github.com/jeantimex/genetic-algorithm/commit/2df4d2ca23ddb86cf28594c632d1b84c628540da))
* optimize the code phase 3 ([88ea1b4](https://github.com/jeantimex/genetic-algorithm/commit/88ea1b4f71bfc1f04e2fe145e44b76aba7809c20))
* optimize the code phase 4 ([a5a31d1](https://github.com/jeantimex/genetic-algorithm/commit/a5a31d1a86aedaf2b109772d2cb36aa2ca71d797))

## [1.0.4](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.3...v1.0.4) (2020-08-08)


### Bug Fixes

* start to optimize the api ([78e46c2](https://github.com/jeantimex/genetic-algorithm/commit/78e46c21e42662b6c2be54371abaf4f09af70b9d))

## [1.0.3](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.2...v1.0.3) (2020-08-07)


### Bug Fixes

* rename library name to genetic ([b37ff75](https://github.com/jeantimex/genetic-algorithm/commit/b37ff7502f6a098dbd61e50b91bfcae97e8a5ddf))

## [1.0.2](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.1...v1.0.2) (2020-08-07)


### Bug Fixes

* basically it is working ([fa859a5](https://github.com/jeantimex/genetic-algorithm/commit/fa859a53f65b94333af171b73738b88d7d1de81f))

## [1.0.1](https://github.com/jeantimex/genetic-algorithm/compare/v1.0.0...v1.0.1) (2020-08-04)


### Bug Fixes

* remove .DS_Store file ([a4b714c](https://github.com/jeantimex/genetic-algorithm/commit/a4b714cef13c41855c010116f2b23979b3317850))

# 1.0.0 (2020-08-04)


### Features

* genetic algorithm initial setup ([e65ab46](https://github.com/jeantimex/genetic-algorithm/commit/e65ab462ab4801ab0ce44c2170be9545b129d23a))


src/index.js
import { clone, maximize, minimize, tournamentSelection } from './utils';

const SelectionStrategy = {
  Tournament: 'Tournament',
  Random: 'Random',
  Rank: 'Rank',
};

const OptimizeStrategy = {
  Maximize: 'Maximize',
  Minimize: 'Minimize',
};

/**
 * Genetic Algorithm
 */
class GeneticAlgorithm {
  constructor(options = {}) {
    const { optimizeStrategy, selectionStrategy, seeds } = options;

    this.optimizeStrategy = optimizeStrategy || OptimizeStrategy.Maximize;
    this.selectionStrategy = selectionStrategy || SelectionStrategy.Tournament;

    this.fitness = null;
    this.mutate = null;
    this.crossover = null;
    this.onEvolve = null;

    this.crossoverRate = 0.3;
    this.mutationRate = 0.3;

    this.iterations = 4000;
    this.population = [];

    this.surviveFittest = true;

    this.isEvolutionCompleted = () => false;
    this.onEvolve = () => {};
  }

  get optimize() {
    if (this.optimizeStrategy === OptimizeStrategy.Maximize) {
      return maximize;
    }
    if (this.optimizeStrategy === OptimizeStrategy.Minimize) {
      return minimize;
    }
  }

  get selectSingle() {
    if (this.selectionStrategy === SelectionStrategy.Tournament) {
      return tournamentSelection(this.optimize);
    }
  }

  get selectParents() {
    if (this.selectionStrategy === SelectionStrategy.Tournament) {
      return (pop) => [this.selectSingle(pop), this.selectSingle(pop)];
    }
  }

  sortPopulation(population) {
    population.sort((a, b) => (this.optimize(a.fitness, b.fitness) ? -1 : 1));
    return population;
  }

  newEntity(chromosome) {
    return { chromosome, fitness: this.fitness(chromosome) };
  }

  evolve() {
    const mutateOrNot = (chromosome) => {
      // applies mutation based on mutation probability
      return Math.random() <= this.mutationRate && this.mutate
        ? this.mutate(clone(chromosome))
        : chromosome;
    };

    // crossover and mutate
    const newPopulation = [];

    if (this.surviveFittest)
      // lets the best solution fall through
      newPopulation.push(this.population[0]);

    while (newPopulation.length < this.population.length) {
      if (
        this.crossoverRate && // if there is a crossover function
        Math.random() <= this.crossoverRate && // base crossover on specified probability
        newPopulation.length + 1 < this.population.length // keeps us from going 1 over the max population size
      ) {
        const [mother, father] = this.selectParents(this.population);
        const children = this.crossover(clone(mother), clone(father))
          .map(mutateOrNot)
          .map(this.newEntity, this);
        newPopulation.push(...children);
      } else {
        newPopulation.push(
          this.newEntity(mutateOrNot(this.selectSingle(this.population)))
        );
      }
    }

    this.population = this.sortPopulation(newPopulation);
  }

  start(seeds) {
    // Create the first population
    this.population = this.sortPopulation(seeds.map(this.newEntity, this));

    for (let generation = 0; generation < this.iterations; generation++) {
      this.onEvolve(generation, this.population);

      if (this.isEvolutionCompleted(this.population)) break;

      this.evolve();
    }
  }
}

export { GeneticAlgorithm, OptimizeStrategy, SelectionStrategy };


src/utils.js
const maximize = (a, b) => a >= b;
const minimize = (a, b) => a < b;

/**
 * Make a deep copy of an object
 *
 * @param {Object} obj
 * @returns {Object}
 */
const clone = (obj) => {
  if (obj == null || typeof obj !== 'object') {
    return obj;
  }
  return JSON.parse(JSON.stringify(obj));
};

const encode = (obj) =>
  JSON.stringify(obj, function (key, value) {
    if (value instanceof Function || typeof value == 'function') {
      return '__func__:' + value.toString();
    }
    if (value instanceof RegExp) {
      return '__regex__:' + value;
    }
    return value;
  });

const decode = (str) =>
  JSON.parse(str, function (key, value) {
    if (typeof value != 'string') {
      return value;
    }
    if (value.lastIndexOf('__func__:', 0) === 0) {
      return eval('(' + value.slice(9) + ')');
    }
    if (value.lastIndexOf('__regex__:', 0) === 0) {
      return eval('(' + value.slice(10) + ')');
    }
    return value;
  });

const tournamentSelection = (optimize) => (population) => {
  const n = population.length;
  const a = population[Math.floor(Math.random() * n)];
  const b = population[Math.floor(Math.random() * n)];
  return optimize(a.fitness, b.fitness) ? a.chromosome : b.chromosome;
};

export { clone, encode, decode, maximize, minimize, tournamentSelection };


examples/browser/example.js
const { GeneticAlgorithm } = genetic;

const ga = new GeneticAlgorithm();

let solution = '';
let last = '';

ga.mutate = function (chromosome) {
  function replaceAt(str, index, character) {
    return (
      str.substr(0, index) + character + str.substr(index + character.length)
    );
  }

  // chromosomal drift
  const i = Math.floor(Math.random() * chromosome.length);
  return replaceAt(
    chromosome,
    i,
    String.fromCharCode(
      chromosome.charCodeAt(i) + (Math.floor(Math.random() * 2) ? 1 : -1)
    )
  );
};

ga.crossover = function (mother, father) {
  // two-point crossover
  const len = mother.length;
  let ca = Math.floor(Math.random() * len);
  let cb = Math.floor(Math.random() * len);
  if (ca > cb) {
    const tmp = cb;
    cb = ca;
    ca = tmp;
  }

  const son =
    father.substr(0, ca) + mother.substr(ca, cb - ca) + father.substr(cb);
  const daughter =
    mother.substr(0, ca) + father.substr(ca, cb - ca) + mother.substr(cb);

  return [son, daughter];
};

ga.fitness = function (chromosome) {
  let fitness = 0;

  for (let i = 0; i < chromosome.length; ++i) {
    // increase fitness for each character that matches
    if (chromosome[i] == solution[i]) fitness += 1;

    // award fractions of a point as we get warmer
    fitness +=
      (127 - Math.abs(chromosome.charCodeAt(i) - solution.charCodeAt(i))) / 50;
  }

  return fitness;
};

ga.isEvolutionCompleted = function (population) {
  // stop running once we've reached the solution
  return population[0].chromosome === solution;
};

ga.onEvolve = function (generation, population) {
  function lerp(a, b, p) {
    return a + (b - a) * p;
  }

  const value = population[0].chromosome;
  last = last || value;

  if (generation !== 0 && value === last) return;

  const solution = [];
  for (let i = 0; i < value.length; ++i) {
    const diff = value.charCodeAt(i) - last.charCodeAt(i);
    let style = 'background: transparent;';
    if (diff > 0) {
      style = 'background: rgb(0,200,50); color: #fff;';
    } else if (diff < 0) {
      style = 'background: rgb(0,100,50); color: #fff;';
    }

    solution.push('<span style="' + style + '">' + value[i] + '</span>');
  }

  let buf = '';
  buf += '<tr>';
  buf += '<td>' + generation + '</td>';
  buf += '<td>' + population[0].fitness.toPrecision(5) + '</td>';
  buf += '<td>' + solution.join('') + '</td>';
  buf += '</tr>';

  const tbody = document.querySelector('#results tbody');
  tbody.innerHTML = buf + tbody.innerHTML;

  last = value;
};

const btnSolve = document.querySelector('#solve');
btnSolve.addEventListener('click', () => {
  const tbody = document.querySelector('#results tbody');
  tbody.innerHTML = '';

  const quote = document.querySelector('#quote');
  solution = quote.value;

  function randomString(len) {
    let text = '';
    const charset = 'abcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < len; i++)
      text += charset.charAt(Math.floor(Math.random() * charset.length));

    return text;
  }

  const populationSize = 250;
  const seeds = [];
  for (let i = 0; i < populationSize; i++) {
    // create random strings that are equal in length to solution
    seeds.push(randomString(solution.length));
  }

  last = '';
  ga.start(seeds);
});


examples/browser/genetic-algorithm.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["genetic"] = factory();
	else
		root["genetic"] = factory();
})(this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = "./src/index.js");
/******/ })
/************************************************************************/
/******/ ({

/***/ "./src/index.js":
/*!**********************!*\
  !*** ./src/index.js ***!
  \**********************/
/*! exports provided: GeneticAlgorithm, OptimizeStrategy, SelectionStrategy */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
eval("__webpack_require__.r(__webpack_exports__);\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"GeneticAlgorithm\", function() { return GeneticAlgorithm; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"OptimizeStrategy\", function() { return OptimizeStrategy; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"SelectionStrategy\", function() { return SelectionStrategy; });\n/* harmony import */ var _utils__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./utils */ \"./src/utils.js\");\nfunction _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\"Cannot call a class as a function\"); } }\n\nfunction _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\"value\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }\n\nfunction _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }\n\n\nvar SelectionStrategy = {\n  Tournament: 'Tournament',\n  Random: 'Random',\n  Rank: 'Rank'\n};\nvar OptimizeStrategy = {\n  Maximize: 'Maximize',\n  Minimize: 'Minimize'\n};\n/**\n * Genetic Algorithm\n */\n\nvar GeneticAlgorithm = /*#__PURE__*/function () {\n  function GeneticAlgorithm() {\n    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};\n\n    _classCallCheck(this, GeneticAlgorithm);\n\n    var optimizeStrategy = options.optimizeStrategy,\n        selectionStrategy = options.selectionStrategy,\n        seeds = options.seeds;\n    this.optimizeStrategy = optimizeStrategy || OptimizeStrategy.Maximize;\n    this.selectionStrategy = selectionStrategy || SelectionStrategy.Tournament;\n    this.fitness = null;\n    this.mutate = null;\n    this.crossover = null;\n    this.generation = null;\n    this.onEvolve = null;\n    this.crossoverRate = 0.3;\n    this.mutationRate = 0.3;\n    this.iterations = 4000;\n    this.fittestAlwaysSurvives = true;\n    this.population = [];\n  }\n\n  _createClass(GeneticAlgorithm, [{\n    key: \"evolve\",\n    value: function evolve(generation) {\n      var _this = this;\n\n      // score and sort by fitness\n      var entities = this.population.map(function (entity) {\n        return {\n          entity: entity,\n          fitness: _this.fitness(entity)\n        };\n      }).sort(function (a, b) {\n        // Put the fittest one in front of the array.\n        return _this.optimize(a.fitness, b.fitness) ? -1 : 1;\n      }); // generation notification\n\n      var mean = entities.reduce(function (a, b) {\n        return a + b.fitness;\n      }, 0) / this.populationSize;\n      var stdev = Math.sqrt(entities.map(function (a) {\n        return (a.fitness - mean) * (a.fitness - mean);\n      }).reduce(function (a, b) {\n        return a + b;\n      }, 0) / this.populationSize);\n      var stats = {\n        maximum: entities[0].fitness,\n        minimum: entities[this.populationSize - 1].fitness,\n        mean: mean,\n        stdev: stdev\n      };\n      var r = this.isFinished ? this.isFinished(entities, generation, stats) : false;\n      var isFinished = !!r || generation == this.iterations - 1;\n      this.onEvolve(entities, generation, stats, isFinished);\n      if (isFinished) return true;\n\n      var mutateOrNot = function mutateOrNot(entity) {\n        // applies mutation based on mutation probability\n        return Math.random() <= _this.mutationRate && _this.mutate ? _this.mutate(Object(_utils__WEBPACK_IMPORTED_MODULE_0__[\"clone\"])(entity)) : entity;\n      }; // crossover and mutate\n\n\n      var newPopulation = [];\n      if (this.fittestAlwaysSurvives) // lets the best solution fall through\n        newPopulation.push(entities[0].entity);\n\n      while (newPopulation.length < this.populationSize) {\n        if (this.crossoverRate && // if there is a crossover function\n        Math.random() <= this.crossoverRate && // base crossover on specified probability\n        newPopulation.length + 1 < this.populationSize // keeps us from going 1 over the max population size\n        ) {\n            var parents = this.selectParents(entities);\n            var children = this.crossover(Object(_utils__WEBPACK_IMPORTED_MODULE_0__[\"clone\"])(parents[0]), Object(_utils__WEBPACK_IMPORTED_MODULE_0__[\"clone\"])(parents[1])).map(mutateOrNot);\n            newPopulation.push(children[0], children[1]);\n          } else {\n          newPopulation.push(mutateOrNot(this.selectSingle(entities)));\n        }\n      }\n\n      this.population = newPopulation;\n    }\n  }, {\n    key: \"start\",\n    value: function start(seeds) {\n      // seed the population\n      this.population = seeds;\n\n      for (var i = 0; i < this.iterations; ++i) {\n        if (this.evolve(i)) {\n          break;\n        }\n      }\n    }\n  }, {\n    key: \"optimize\",\n    get: function get() {\n      if (this.optimizeStrategy === OptimizeStrategy.Maximize) {\n        return _utils__WEBPACK_IMPORTED_MODULE_0__[\"maximize\"];\n      }\n\n      if (this.optimizeStrategy === OptimizeStrategy.Minimize) {\n        return _utils__WEBPACK_IMPORTED_MODULE_0__[\"minimize\"];\n      }\n    }\n  }, {\n    key: \"selectSingle\",\n    get: function get() {\n      if (this.selectionStrategy === SelectionStrategy.Tournament) {\n        return Object(_utils__WEBPACK_IMPORTED_MODULE_0__[\"tournamentSelection\"])(this.optimize);\n      }\n    }\n  }, {\n    key: \"selectParents\",\n    get: function get() {\n      var _this2 = this;\n\n      if (this.selectionStrategy === SelectionStrategy.Tournament) {\n        return function (pop) {\n          return [_this2.selectSingle(pop), _this2.selectSingle(pop)];\n        };\n      }\n    }\n  }, {\n    key: \"populationSize\",\n    get: function get() {\n      return this.population.length;\n    }\n  }]);\n\n  return GeneticAlgorithm;\n}();\n\n\n\n//# sourceURL=webpack://genetic/./src/index.js?");

/***/ }),

/***/ "./src/utils.js":
/*!**********************!*\
  !*** ./src/utils.js ***!
  \**********************/
/*! exports provided: clone, encode, decode, maximize, minimize, tournamentSelection */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
eval("__webpack_require__.r(__webpack_exports__);\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"clone\", function() { return clone; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"encode\", function() { return encode; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"decode\", function() { return decode; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"maximize\", function() { return maximize; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"minimize\", function() { return minimize; });\n/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \"tournamentSelection\", function() { return tournamentSelection; });\nfunction _typeof(obj) { \"@babel/helpers - typeof\"; if (typeof Symbol === \"function\" && typeof Symbol.iterator === \"symbol\") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === \"function\" && obj.constructor === Symbol && obj !== Symbol.prototype ? \"symbol\" : typeof obj; }; } return _typeof(obj); }\n\nvar maximize = function maximize(a, b) {\n  return a >= b;\n};\n\nvar minimize = function minimize(a, b) {\n  return a < b;\n};\n/**\n * Make a deep copy of an object\n *\n * @param {Object} obj\n * @returns {Object}\n */\n\n\nvar clone = function clone(obj) {\n  if (obj == null || _typeof(obj) !== 'object') {\n    return obj;\n  }\n\n  return JSON.parse(JSON.stringify(obj));\n};\n\nvar encode = function encode(obj) {\n  return JSON.stringify(obj, function (key, value) {\n    if (value instanceof Function || typeof value == 'function') {\n      return '__func__:' + value.toString();\n    }\n\n    if (value instanceof RegExp) {\n      return '__regex__:' + value;\n    }\n\n    return value;\n  });\n};\n\nvar decode = function decode(str) {\n  return JSON.parse(str, function (key, value) {\n    if (typeof value != 'string') {\n      return value;\n    }\n\n    if (value.lastIndexOf('__func__:', 0) === 0) {\n      return eval('(' + value.slice(9) + ')');\n    }\n\n    if (value.lastIndexOf('__regex__:', 0) === 0) {\n      return eval('(' + value.slice(10) + ')');\n    }\n\n    return value;\n  });\n};\n\nvar tournamentSelection = function tournamentSelection(optimize) {\n  return function (pop) {\n    var n = pop.length;\n    var a = pop[Math.floor(Math.random() * n)];\n    var b = pop[Math.floor(Math.random() * n)];\n    return optimize(a.fitness, b.fitness) ? a.entity : b.entity;\n  };\n};\n\n\n\n//# sourceURL=webpack://genetic/./src/utils.js?");

/***/ })

/******/ });
});